tox-offline-draft.md

```diff
@@ -5,77 +5,154 @@
+
+## Packet encryption
+
+There are 2 layers of encryption for a message, first the SENDER encrypts the
+message using the agreed/shared key. (TODO: Is this the Tox per-friend key we
+already use for messaging, or is this OM specific?) 
+
+The SENDER then encrypts that message for us using the SENDER -> RECEIVER
+shared key. That way the STORAGE_NODE can receive the message, encrypt it,
+and write it to disk forgetting it exists until requested. Then once
+requested, it'll send the already encrypted message to the reciever, and
+delete the data on disk.  Invalidate the current POW, and the whole cycle
+will start over again.
+
+```
+SENDER -> DHT_Node [The Message]
+    encrypts to friend       --> [SENDER : RECEIVER  key_pair || The Message]
+    encrypts to storage_node --> [sender:DHT_node shared key [SENDER : RECEIVER  key_pair || The Message]]
+DHT_node  
+    decrypts from sender     --> [sender:friend shared key || The Message]
+    encrypts to friend       --> [DHT_node shared key [SENDER : RECEIVER  key_pair || The Message]]
+RECEIVER [sender:DHT_node shared key 
+    decrypts the message from the node   --> [SENDER : RECEIVER  key_pair || The Message]
+    decrypts the message from the friend --> [The Message]
+RECEIVER DONE --> display [The Message]
+```
 
 ## PoW function
 
-Something that can be adjusted, look into BitMessage.
+PoW job is computed on STORAGE_NODE (low cost) then sent to either SENDER or
+RECEIVER to compute for proof (high cost).
+
+Looking at BitMessage's implementation [1], SHA512 is used.
+
+Their [function for PoW difficulty](
+https://bitmessage.org/wiki/File:POW_forumla_2.png ) looks substantial.  The
+payloadLengthExtraBytes var is to give weight to smaller messages, garbage
+bytes are not actually appended to message packets.
 
 # References
 
-[0] https://morph.is/v0.8/dpush-whitepaper.odt
+[0] [Dpush -- A scalable decentralized spam resistant unsolicited messaging protocol](dpush.md)
+    New implementation is not really similar to Dpush anymore, but fun read nonetheless.
+    
+[1] [BitMessage Spec v3 PoW Summary](https://bitmessage.org/wiki/POW) 
+
+
+## TODO
+    Consider or reject
+        allowing friends to store messages for shared friends (maybe with a lower proof of work)
+        multi-device compatability
+        allowing senders to increase POW space
+        Define and adapt use case to systems where tox may get killed trying to compute POW (android, iOS, etc.)
+            Persistant notification keep processes alive, no? Android may still kill them if resources are low though.
+            yes, but doze dosen't like you using the CPU that much -- understandable
+            And iOS kills them after ~10m iirc.
+                @dvor would know the answer to this. Also if you get push messages, and the corresponding wakeup from iOS, and then use a lot of network, or battery life. iOS will just stop sending you messages.
```

Comment:
@dvor the storage space is allocated per friend. Just by knowing who the message is from. You already know where you need to go looking.